vkrunner/vr-pipeline.c

```diff
@@ -709,30 +729,105 @@ create_vk_descriptor_set_layout(struct vr_pipeline *pipeline,
                 bindings[i].descriptorType = descriptor_type;
                 bindings[i].descriptorCount = 1;
                 bindings[i].stageFlags = pipeline->stages;
+
+                if (prev_desc_set != buffer->desc_set) {
+                        info[n_desc_sets].n_bindings =
+                                &bindings[i] - info[n_desc_sets].bindings;
+                        ++n_desc_sets;
+                        info[n_desc_sets].desc_set = buffer->desc_set;
+                        info[n_desc_sets].bindings = &bindings[i];
+                        prev_desc_set = buffer->desc_set;
+                }
         }
+        info[n_desc_sets].n_bindings =
+                &bindings[n_buffers] - info[n_desc_sets].bindings;
+        ++n_desc_sets;
 
-        VkDescriptorSetLayoutCreateInfo create_info = {
-                .sType = VK_STRUCTURE_TYPE_DESCRIPTOR_SET_LAYOUT_CREATE_INFO,
-                .bindingCount = n_buffers,
-                .pBindings = bindings
+        VkDescriptorPoolSize pool_size_candidates[] = {
+                {
+                        .type = VK_DESCRIPTOR_TYPE_UNIFORM_BUFFER,
+                        .descriptorCount = n_ubo
+                },
+                {
+                        .type = VK_DESCRIPTOR_TYPE_STORAGE_BUFFER,
+                        .descriptorCount = n_ssbo
+                },
         };
 
-        VkDescriptorSetLayout descriptor_set_layout;
-
-        res = vkfn->vkCreateDescriptorSetLayout(pipeline->window->device,
-                                                &create_info,
-                                                NULL, /* allocator */
-                                                &descriptor_set_layout);
+        VkDescriptorPoolSize *pool_sizes;
+        uint32_t n_pool_sizes;
+        if (!n_ubo && !n_ssbo) {
+                vr_fatal("Both n_ubo and n_ssbo are zeros");
+        } else if (!n_ubo) {
+                pool_sizes = &pool_size_candidates[1];
+                n_pool_sizes = 1;
+        } else if (!n_ssbo) {
+                pool_sizes = &pool_size_candidates[0];
+                n_pool_sizes = 1;
+        } else {
+                pool_sizes = pool_size_candidates;
+                n_pool_sizes = 2;
+        }
+
```

Comment:
This construct seems a little odd. Maybe you could just have `pool_sizes[2]` and then do something like:

``` C
if (n_ubo) {
        pool_sizes[n_pool_sizes].type = ubo;
        pool_sizes[n_pool_sizes].descriptorCount = n_ubo;
        n_pool_sizes++;
}
/* and then the same for n_ssbos */
```